File: app/src/main/java/com/chess/cryptobot/view/MainFragment.java
package com.chess.cryptobot.view;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.chess.cryptobot.R;
import com.chess.cryptobot.content.ContextHolder;
import com.chess.cryptobot.model.ViewItem;
import com.chess.cryptobot.view.adapter.RecyclerViewAdapter;

public abstract class MainFragment<T extends RecyclerView.ViewHolder> extends Fragment implements SwipeRefreshLayout.OnRefreshListener {
    private RecyclerViewAdapter<T> adapter;
    private ContextHolder holder;
    private SwipeRefreshLayout swipeRefreshLayout;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = initFragmentView(inflater, container);
        holder = initHolder();

        RecyclerView recyclerView = initRecyclerView(view);
        adapter = initAdapter(holder);
        recyclerView.setAdapter(adapter);
        LinearLayoutManager layoutManager = new LinearLayoutManager(getActivity());
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setHasFixedSize(true);

        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(recyclerView.getContext(),
                layoutManager.getOrientation());
        dividerItemDecoration.setDrawable(getResources().getDrawable(R.drawable.divider, null));
        recyclerView.addItemDecoration(dividerItemDecoration);

        swipeRefreshLayout = initSwipeRefresh(view);
        swipeRefreshLayout.setOnRefreshListener(this);
        swipeRefreshLayout.setColorSchemeResources(R.color.colorSecondary, R.color.colorSecondaryDark);

        return view;
    }

    @Override
    public void onRefresh() {
        beforeRefresh();
        swipeRefreshLayout.setRefreshing(false);
        holder.updateAllItems();
    }

    protected abstract void beforeRefresh();

    protected abstract View initFragmentView(LayoutInflater inflater, ViewGroup container);

    protected abstract ContextHolder initHolder();

    protected abstract RecyclerView initRecyclerView(View view);

    protected abstract RecyclerViewAdapter<T> initAdapter(ContextHolder holder);

    protected abstract SwipeRefreshLayout initSwipeRefresh(View view);

    public void addItem() {
        adapter.notifyItemInserted();
    }

    public void updateItem(ViewItem item) {
        adapter.updateItem(item);
    }

    public void deleteItemByPosition(int position) {
        adapter.deleteItem(position);
    }

    public String itemNameByPosition(int position) {
        return adapter.itemNameByPosition(position);
    }

    public void updateAllItems() {
        adapter.notifyDataSetChanged();
    }

    public void makeToast(String message) {
        if (message != null && !message.isEmpty()) {
            Toast.makeText(getActivity(), message, Toast.LENGTH_LONG).show();
        }
    }

    public void showSpinner() {
        FragmentActivity activity = getActivity();
        if (activity == null) return;
        ProgressBar spinner = activity.findViewById(R.id.progressBar);
        spinner.setVisibility(View.VISIBLE);
    }

    public void hideSpinner() {
        FragmentActivity activity = getActivity();
        if (activity == null) return;
        ProgressBar spinner = activity.findViewById(R.id.progressBar);
        spinner.setVisibility(View.GONE);
    }

    ContextHolder getHolder() {
        return holder;
    }
}

File: app/src/main/java/com/chess/cryptobot/view/MainFragment.kt
package com.chess.cryptobot.view

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import android.widget.Toast
import androidx.core.content.res.ResourcesCompat
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout.OnRefreshListener
import com.chess.cryptobot.R
import com.chess.cryptobot.content.ContextHolder
import com.chess.cryptobot.model.ViewItem
import com.chess.cryptobot.view.adapter.RecyclerViewAdapter

abstract class MainFragment<T : RecyclerView.ViewHolder> : Fragment(), OnRefreshListener {
    private var adapter: RecyclerViewAdapter<T>? = null
    lateinit var holder: ContextHolder
        private set

    private var swipeRefreshLayout: SwipeRefreshLayout? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = initFragmentView(inflater, container)
        holder = initHolder()
        val recyclerView = initRecyclerView(view)
        adapter = initAdapter(holder)
        recyclerView.adapter = adapter
        val layoutManager = LinearLayoutManager(activity)
        recyclerView.layoutManager = layoutManager
        recyclerView.setHasFixedSize(true)
        val dividerItemDecoration = DividerItemDecoration(recyclerView.context,
                layoutManager.orientation)
        dividerItemDecoration.setDrawable(ResourcesCompat.getDrawable(resources, R.drawable.divider, null)!!)
        recyclerView.addItemDecoration(dividerItemDecoration)
        swipeRefreshLayout = initSwipeRefresh(view)
        swipeRefreshLayout!!.setOnRefreshListener(this)
        swipeRefreshLayout!!.setColorSchemeResources(R.color.colorSecondary, R.color.colorSecondaryDark)
        return view
    }

    override fun onRefresh() {
        beforeRefresh()
        swipeRefreshLayout!!.isRefreshing = false
        holder.updateAllItems()
    }

    protected abstract fun beforeRefresh()
    protected abstract fun initFragmentView(inflater: LayoutInflater, container: ViewGroup?): View
    protected abstract fun initHolder(): ContextHolder
    protected abstract fun initRecyclerView(view: View): RecyclerView
    protected abstract fun initAdapter(holder: ContextHolder?): RecyclerViewAdapter<T>
    protected abstract fun initSwipeRefresh(view: View): SwipeRefreshLayout
    fun addItem() {
        adapter!!.notifyItemInserted()
    }

    fun updateItem(item: ViewItem?) {
        adapter!!.updateItem(item)
    }

    fun deleteItemByPosition(position: Int) {
        adapter!!.deleteItem(position)
    }

    fun itemNameByPosition(position: Int): String {
        return adapter!!.itemNameByPosition(position)
    }

    fun updateAllItems() {
        adapter!!.notifyDataSetChanged()
    }

    fun makeToast(message: String?) {
        if (message != null && message.isNotEmpty()) {
            Toast.makeText(activity, message, Toast.LENGTH_LONG).show()
        }
    }

    fun showSpinner() {
        val activity = activity ?: return
        val spinner = activity.findViewById<ProgressBar>(R.id.progressBar)
        spinner.visibility = View.VISIBLE
    }

    fun hideSpinner() {
        val activity = activity ?: return
        val spinner = activity.findViewById<ProgressBar>(R.id.progressBar)
        spinner.visibility = View.GONE
    }

}